parser/src/format/character_separated/w3c_extended_log.rs

```diff
@@ -0,0 +1,181 @@
+    MissingFields,
+    #[error("file headers are malformed")]
+    InvalidHeader,
+
+    #[error("invalid encoding of log file")]
+    InvalidEncoding(#[from] std::str::Utf8Error),
+
+    #[error("reading input: {0}")]
+    Io(#[from] std::io::Error),
+}
+
+pub fn new_w3c_extended_log_parser() -> Box<dyn Parser> {
+    Box::new(W3cLogParser)
+}
+pub struct W3cLogParser;
+
+/// A parsed directive from the header of a log file.
+#[allow(dead_code)]
+struct Directive {
+    name: String,
+    value: String,
+}
+
+/// Represents the parsed header of a W3C extended log file, which is a series of directives.
+/// The `#Fields` directive is the only one that's significant to parsing the rest of the file,
+/// but all the other directives are preserved here so that they can be added to the output
+/// documents if desired.
+struct Header {
+    fields: Vec<String>,
+    other_directives: Vec<Directive>,
+}
+
+impl W3cLogParser {
+    fn parse_header(input: &[u8]) -> Result<Header, Error> {
+        let mut header = Header {
+            fields: Vec::new(),
+            other_directives: Vec::new(),
+        };
+        for result in input.lines() {
+            let line = result?;
+            if !line.starts_with("#") {
+                break;
+            }
+            let line = &line[1..];
+            let (name, rem) = line.split_once(':').ok_or_else(|| Error::InvalidHeader)?;
+            let name = name.trim();
+            tracing::debug!(%name, value = %rem, "parsed w3c extended log directive");
+            if name == "Fields" {
+                header.fields = rem.split_ascii_whitespace().map(String::from).collect();
+            } else {
+                header.other_directives.push(Directive {
+                    name: name.to_string(),
+                    value: rem.trim().to_string(),
+                })
+            }
+        }
+        if header.fields.is_empty() {
+            Err(Error::MissingFields)
+        } else {
+            Ok(header)
+        }
+    }
+}
+
+impl Parser for W3cLogParser {
+    fn parse(&self, config: &ParseConfig, content: Input) -> Result<Output, ParseError> {
+        // We'll peek at a up to 8KiB in order to try to parse the header. The Fields directive is
+        // the only thing we really _need_ from the header, and that's always supposed to be the
+        // second directive (right after Version). So practically, this should always be enough,
+        // even if there's hundreds of columns.
+        let (prefix, input) = content.peek(8192)?;
+        let header =
+            W3cLogParser::parse_header(&prefix).map_err(|e| ParseError::Parse(Box::new(e)))?;
+        // TODO: (optionally) add the extra directives from header as fields to each json object.
+
+        let projections = build_projections(config)?;
+        let headers = resolve_headers(header.fields, projections, &["-"]);
+
+        let reader = csv::ReaderBuilder::new()
+            .delimiter(b'\t')
+            .comment(Some(b'#'))
```

Comment:
This is clever (by telling the csv parser that header rows are comments, we don't have to worry about skipping ahead in the input stream manually), maybe worthy of a comment.

Reply:
good call, I added a comment.